backend/x11/backend.c

```diff
@@ -560,31 +597,65 @@ struct wlr_backend *wlr_x11_backend_create(struct wl_display *display,
 	xcb_create_colormap(x11->xcb, XCB_COLORMAP_ALLOC_NONE, x11->colormap,
 		x11->screen->root, x11->visualid);
 
-	// DRI3 may return a render node (Xwayland) or an authenticated primary
-	// node (plain Glamor).
-	x11->drm_fd = query_dri3_drm_fd(x11);
-	if (x11->drm_fd < 0) {
-		wlr_log(WLR_ERROR, "Failed to query DRI3 DRM FD");
-		goto error_event;
+	if (!query_formats(x11)) {
+		wlr_log(WLR_ERROR, "Failed to query supported DRM formats");
+		return false;
 	}
 
-	char *drm_name = drmGetDeviceNameFromFd2(x11->drm_fd);
-	wlr_log(WLR_DEBUG, "Using DRM node %s", drm_name);
-	free(drm_name);
+	const struct wlr_drm_format *pixmap_format = NULL;
```

Comment:
Instead of a `struct wlr_drm_format`, we can declare a `struct wlr_drm_format_set` variable here, set it to either `dri3_formats` or `shm_formats` in each `if` branch, and de-duplicate the `wlr_drm_format_set_get` call. Does that make sense?